Installation issue: py-numpy with GCC 4.8.5

### Steps to reproduce the issue

```console
$ s spec -I py-numpy
Input spec
--------------------------------
 -   py-numpy

Concretized
--------------------------------
 -   py-numpy@1.24.2%gcc@4.8.5+blas+lapack build_system=python_pip patches=873745d arch=linux-rhel7-sandybridge
[+]      ^openblas@0.3.21%gcc@4.8.5~bignuma~consistent_fpcsr+fortran~ilp64+locking+pic+shared build_system=makefile patches=114f95f,a4c642f,c20f518,d3d9b15 symbol_suffix=none threads=none arch=linux-rhel7-sandybridge
[+]          ^perl@5.36.0%gcc@4.8.5+cpanm+open+shared+threads build_system=generic arch=linux-rhel7-sandybridge
[+]              ^berkeley-db@18.1.40%gcc@4.8.5+cxx~docs+stl build_system=autotools patches=26090f4,b231fcc arch=linux-rhel7-sandybridge
[+]      ^py-cython@0.29.33%gcc@4.8.5 build_system=python_pip patches=71de066 arch=linux-rhel7-sandybridge
[+]      ^py-pip@23.0%gcc@4.8.5 build_system=generic arch=linux-rhel7-sandybridge
[+]      ^py-setuptools@63.0.0%gcc@4.8.5 build_system=generic arch=linux-rhel7-sandybridge
[+]      ^py-wheel@0.37.1%gcc@4.8.5 build_system=generic arch=linux-rhel7-sandybridge
[+]      ^python@3.10.10%gcc@4.8.5+bz2+crypt+ctypes+dbm~debug+libxml2+lzma~nis+optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tkinter+uuid+zlib build_system=generic patches=0d98e93,7d40923,f2fd060 arch=linux-rhel7-sandybridge
[+]          ^bzip2@1.0.8%gcc@4.8.5~debug~pic+shared build_system=generic arch=linux-rhel7-sandybridge
[+]              ^diffutils@3.9%gcc@4.8.5 build_system=autotools arch=linux-rhel7-sandybridge
[+]          ^expat@2.5.0%gcc@4.8.5+libbsd build_system=autotools arch=linux-rhel7-sandybridge
[+]              ^libbsd@0.11.7%gcc@4.8.5 build_system=autotools arch=linux-rhel7-sandybridge
[+]                  ^libmd@1.0.4%gcc@4.8.5 build_system=autotools arch=linux-rhel7-sandybridge
[+]          ^gdbm@1.23%gcc@4.8.5 build_system=autotools arch=linux-rhel7-sandybridge
[+]          ^gettext@0.21.1%gcc@4.8.5+bzip2+curses+git~libunistring+libxml2+tar+xz build_system=autotools arch=linux-rhel7-sandybridge
[+]              ^libiconv@1.17%gcc@4.8.5 build_system=autotools libs=shared,static arch=linux-rhel7-sandybridge
[+]              ^libxml2@2.10.3%gcc@4.8.5~python build_system=autotools arch=linux-rhel7-sandybridge
[+]              ^tar@1.34%gcc@4.8.5 build_system=autotools zip=pigz arch=linux-rhel7-sandybridge
[+]                  ^pigz@2.7%gcc@4.8.5 build_system=makefile arch=linux-rhel7-sandybridge
[+]                  ^zstd@1.5.4%gcc@4.8.5~ipo+programs build_system=cmake build_type=RelWithDebInfo compression=none generator=make libs=shared,static arch=linux-rhel7-sandybridge
[+]                      ^cmake@3.25.2%gcc@4.8.5~doc+ncurses+ownlibs~qt build_system=generic build_type=Release arch=linux-rhel7-sandybridge
[+]                      ^gmake@4.4.1%gcc@4.8.5~guile build_system=autotools arch=linux-rhel7-sandybridge
[+]          ^libffi@3.4.4%gcc@4.8.5 build_system=autotools arch=linux-rhel7-sandybridge
[+]          ^libxcrypt@4.4.33%gcc@4.8.5~obsolete_api build_system=autotools arch=linux-rhel7-sandybridge
[+]          ^ncurses@6.4%gcc@4.8.5~symlinks+termlib abi=none build_system=autotools arch=linux-rhel7-sandybridge
[+]          ^openssl@1.1.1t%gcc@4.8.5~docs~shared build_system=generic certs=mozilla arch=linux-rhel7-sandybridge
[+]              ^ca-certificates-mozilla@2023-01-10%gcc@4.8.5 build_system=generic arch=linux-rhel7-sandybridge
[+]          ^pkgconf@1.8.0%gcc@4.8.5 build_system=autotools arch=linux-rhel7-sandybridge
[+]          ^readline@8.2%gcc@4.8.5 build_system=autotools patches=bbf97f1 arch=linux-rhel7-sandybridge
[+]          ^sqlite@3.40.1%gcc@4.8.5+column_metadata+dynamic_extensions+fts~functions+rtree build_system=autotools arch=linux-rhel7-sandybridge
[+]          ^util-linux-uuid@2.36.2%gcc@4.8.5 build_system=autotools arch=linux-rhel7-sandybridge
[+]          ^xz@5.4.1%gcc@4.8.5~pic build_system=autotools libs=shared,static arch=linux-rhel7-sandybridge
[+]          ^zlib@1.2.13%gcc@4.8.5+optimize+pic+shared build_system=makefile arch=linux-rhel7-sandybridge
```


### Error message

<details><summary>Error message</summary>

<pre>
...
  gcc: error: unrecognized command line option '-mavx512f'
...
  /scratch/users/adamjs5/spack-stage-py-numpy-1.24.2-oc6ad2hgosendojq4rvyidiqlibrecb4/spack-src/numpy/distutils/checks/cpu_avx512f.c:18:5: error: unknown type name '__m512i'
...
  /scratch/users/adamjs5/spack-stage-py-numpy-1.24.2-oc6ad2hgosendojq4rvyidiqlibrecb4/spack-src/numpy/distutils/checks/cpu_avx512f.c:19:5: error: incompatible type for argument 1 of '_mm_cvtsi128_si32'
...
  numpy/core/src/npysort/binsearch.cpp:293:52: error: array must be initialized with a brace-enclosed initializer
...
  numpy/core/src/npysort/binsearch.cpp:293:52: error: too many initializers for 'const std::array<int (*)(const char*, const char*, const char*, char*, long int, long int, long int, long int, long int, long int, tagPyArrayObject*), 2ul>'
</pre></details>


### Information on your system

* **Spack:** 0.20.0.dev0
* **Python:** 3.6.8
* **Platform:** linux-rhel7-broadwell
* **Concretizer:** clingo


### Additional information

* [spack-build-out.txt](https://github.com/spack/spack/files/11255162/spack-build-out.txt)
* [spack-build-env-mods.txt](https://github.com/spack/spack/files/11255163/spack-build-env-mods.txt)

@rgommers does numpy no longer support GCC 4.8?

@alalazo is this a bug in archspec?

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

--- a/var/spack/repos/builtin/packages/py-numpy/package.py
+++ b/var/spack/repos/builtin/packages/py-numpy/package.py
@@ -161,6 +161,7 @@ class PyNumpy(PythonPackage):
     # meson.build
     # https://docs.scipy.org/doc/scipy/dev/toolchain.html#compilers
     conflicts("%gcc@:8.3", when="@1.26:", msg="NumPy requires GCC >= 8.4")
+    conflicts("%gcc@:5", when="@1.24:", msg="NumPy requires GCC >= 6")
     conflicts("%gcc@:4.7", msg="NumPy requires GCC >= 4.8")
     conflicts(
         "%msvc@:19.19",

--- a/var/spack/repos/builtin/packages/py-numpy/package.py
+++ b/var/spack/repos/builtin/packages/py-numpy/package.py
@@ -161,7 +161,7 @@ class PyNumpy(PythonPackage):
     # meson.build
     # https://docs.scipy.org/doc/scipy/dev/toolchain.html#compilers
     conflicts("%gcc@:8.3", when="@1.26:", msg="NumPy requires GCC >= 8.4")
-    conflicts("%gcc@:5", when="@1.24:", msg="NumPy requires GCC >= 6")
+    conflicts("%gcc@:6.4", when="@1.23:", msg="NumPy requires GCC >= 6.5")
     conflicts("%gcc@:4.7", msg="NumPy requires GCC >= 4.8")
     conflicts(
         "%msvc@:19.19",

